dropgz/pkg/embed/payload.go

```diff
@@ -86,6 +87,17 @@ func deploy(src, dest string) error {
 		return err
 	}
 	defer rc.Close()
+	if _, err := os.Stat(dest); err == nil || !os.IsNotExist(err) {
```

Comment:
We need to add a nested if for handling the case where no file exists.
```
	_, err = os.Stat(dest)
	if err != nil {
		if !os.IsNotExist(err) {
			return err
		}
	} else {
		oldDest := dest + oldFileSuffix
		if err = os.RemoveAll(oldDest); err != nil {
			return errors.Wrapf(err, "failed to remove the %s", oldDest)
		}
		if err = os.Rename(dest, oldDest); err != nil {
			return errors.Wrapf(err, "failed to rename the %s to %s", dest, oldDest)
		}
	}
```

Even if the `err` is not `nil`, we don't want the program to fail if the file does not exists. If there are any other reason for failure then it should be fine. Does this make sense ?

Reply:
Okay, I follow now... so the idea is basically:

```go
// check for existence:
// if the path doesn't exist, don't do anything
// if something inexplicably horrible happened return an error
// otherwise do the renaming
```

If that be the case, it might make sense to extract it as a function so that we can short-circuit in the case the path doesn't exist:

```go
func renameExisting(dest string) error {
       // check for the file's existence
       _, err := os.Stat(dest)
       if os.IsNotExist(err) {
              // this is perfectly fine because XXX
              return nil
       }
       if err != nil {
              // something catastrophic happened with Stat so...
              return errors.Wrap(err, "checking for destination path existence")
       }

       // the old file exists, thus we can proceed with renaming it:
       // ...
}
```

Then when you go to use it, you can simply:

```go
err := renameExisting(dest)
if err != nil {
       return errors.Wrap(err, "renaming existing path")
}
```

Reply:
I think I agree with Evan, let me just update the PR. Thanks !